### PR TITLE
FUSE versions view

### PR DESCRIPTION
**Pull request by ThomasWaldmann**
for mixed repos, --prefix would make sense here.
also, not mounting all archives, but just last N would also make sense.
we'll do that after the first/last PR is merged.

### CODE REVIEW

File: src/borg/fuse.py [hunk 1]
```diff
@@ -98,27 +106,48 @@ def process_archive(self, archive, prefix=[]):
...
+                    ident = adler32(chunkid, ident)
+                name = b'%s.%08x' % (name, ident)
+            return name
+
+        if versions_view:
+            parent = self.process_inner(name, parent)
+            name = version_name(name, item)
+        self.process_real_leaf(name, item, parent, prefix)
+
+    def process_real_leaf(self, name, item, parent, prefix):
```
Comment (ThomasWaldmann) [inline src/borg/fuse.py]: this is just the old code moved into a method.

File: src/borg/fuse.py [hunk 2]
```diff
@@ -98,27 +106,48 @@ def process_archive(self, archive, prefix=[]):
...
+            item = self.cache.get(inode)
+            item.nlink = item.get('nlink', 1) + 1
+            self.items[inode] = item
+        else:
+            inode = self.cache.add(item)
+        self.parent[inode] = parent
+        if name:
+            self.contents[parent][name] = inode
+
+    def process_inner(self, name, parent):
```
Comment (ThomasWaldmann) [inline src/borg/fuse.py]: this is just the old code moved into a method.

File: src/borg/fuse.py [hunk 3]
```diff
@@ -98,27 +106,48 @@ def process_archive(self, archive, prefix=[]):
...
+            if 'chunks' in item:
+                ident = 0
+                for chunkid, _, _ in item.chunks:
+                    ident = adler32(chunkid, ident)
+                name = b'%s.%08x' % (name, ident)
+            return name
+
+        if versions_view:
+            parent = self.process_inner(name, parent)
+            name = version_name(name, item)
```
Comment (ThomasWaldmann) [inline src/borg/fuse.py]: these few lines are the whole magic. :)

**Comment by ThomasWaldmann** [comment]
rebased on current master

**Comment by ThomasWaldmann** [comment]
Looks now like this (showing 2 versions of fuse.py coming from 2 archives):
```
$ borg mount repo mnt
$ ls -lR mnt/src/borg/fuse.py
mnt/src/borg/fuse.py:
total 23
-rw-rw-r-- 1 tw tw 12026 Aug 16 02:02 fuse.py.00a40f83
-rw-rw-r-- 1 tw tw 12022 Aug 16 01:48 fuse.py.e3010da6
```
You see there is a directory in place of where usually would be a file.
In that directory, there are the versions of that file.
The filenames are repeated, so it is easier to copy out a file of there and still know its name.
Files have a suffix, computed via adler32(file's chunkids).

**Comment by ThomasWaldmann** [comment]
how about merging this mostly "as is" and declaring it an experimental feature that is subject to change? so we could actually use it and get more feedback from practice.

**Comment by enkore** [comment]
I think the only thing missing to do it would be making `versions_view` a mount option.

**Comment by ThomasWaldmann** [comment]
collapse / merge?

**Comment by enkore** [comment]
lgtm. changelog, collapse, merge.

**Pull request merged**